download-runner.sh

```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env bash
+set -o pipefail
+retryAttempt=1
+retryInterval=30
+retryMaxLimit=10
+version=1.0.0
+
+mkVmDir() {
+  mkdir -p "${HOME}/.tart/vms/${vmImageName}"
+}
+
+mkXcodeDir() {
+  mkdir -p "${HOME}/images"
+}
+
+printVersion() {
+  echo "${version}"
+}
+
+printHelp() {
+  printf '%s\n' "Download, extract, validate, and install macOS VM and Xcode images."
+  printf '%s\n' "By default, latest macOS VM and Xcode image will be installed."
+  printf 'Usage: %s [runner-version] [-h|--help]\n' "$0"
+  printf '\t%s\n' "runner-version: Specify the Appcircle runner version optionally."
+  printf '\t%s\n' "-h, --help: Prints help."
+  printf '\t%s\n' "-v, --version: Prints version."
+}
+
+parseArguments() {
+  args=("$@")
+  for arg in "${args[@]}"; do
+    if [[ $arg == "--help" || $arg == "-h" ]]; then
+      printHelp
+      exit 0
+    elif [[ $arg == "--version" || $arg == "-v" ]]; then
+      printVersion
+      exit 0
+    fi
+  done
+
+  if [[ "$#" -ne 0 ]] && [[ "$#" -ne 1 ]]; then
+    echo "Illegal number of parameters." >&2
+    printHelp >&2
+    exit 1
+  fi
+
+  runnerVersion=$1
+  if [[ -z $runnerVersion ]]; then
+    vmImageName=$(getTheLatestVmImageName)
+    xcodeImageName=$(getTheLatestXcodeImageName)
+  else
+    vmImageName="macOS_${runnerVersion}"
+    xcodeImageName="xcode_${runnerVersion}"
+  fi
+  vmImageFile="$vmImageName.tar.gz"
+  xcodeImageFile="$xcodeImageName.tar.gz"
+
+}
+
+getTheLatestVmImageName() {
+  latestVmImageName=$(getTheLatestFile "macOS*")
+  echo "$latestVmImageName"
+}
+
+getTheLatestXcodeImageName() {
+  latestXcodeImageName=$(getTheLatestFile "xcodes*")
+  echo "$latestXcodeImageName"
+}
+
+getTheLatestFile() {
+  fileToSearch=$1
+  fileList=$(curl -fsSL -X GET "https://storage.googleapis.com/storage/v1/b/appcircle-dev-common/o?matchGlob=self-hosted/${fileToSearch}")
+  latestRunner=$(echo "$fileList" | grep -o '"name": "[^"]*' | tail -n 1)
+  fileName=$(basename "$latestRunner")
+  echo "${fileName%.tar.gz}"
+}
+
+downloadFileFromBucket() {
+  fileToDownload=$1
+  curl -f -L -O -C - "https://storage.googleapis.com/appcircle-dev-common/self-hosted/$fileToDownload"
+  if [[ "$?" != 0 ]]; then
+    if [[ "$retryAttempt" -gt "$retryMaxLimit" ]]; then
+      retryAttempt=$((retryAttempt - 1))
+      echo "Failed to download the VM image in $retryAttempt attempt. Please check your network." >&2
+      exit 1
+    fi
+    echo "Download failed. Re-trying: $retryAttempt."
+    retryAttempt=$((retryAttempt + 1))
+    sleep $retryInterval
+    downloadVmImage
```

Comment:
IMO we have a bug here. Retry is not working correctly on my side.

Recursion should be done with a call to the same command instead of `downloadVmImage()`. This function is not only used for VM images but also for Xcode images. :hourglass_flowing_sand:

Reply:
Fixed function name. 
Also fixed retry attempt.

Reply:
Also moved decreasing the retry counter operation into echo.